Src/PerformanceCollector/Shared/Implementation/QuickPulse/QuickPulseCollectionStateManager.cs

```diff
@@ -0,0 +1,215 @@
+            }
+
+            if (onStopCollection == null)
+            {
+                throw new ArgumentNullException(nameof(onStopCollection));
+            }
+
+            if (onSubmitSamples == null)
+            {
+                throw new ArgumentNullException(nameof(onSubmitSamples));
+            }
+
+            if (onReturnFailedSamples == null)
+            {
+                throw new ArgumentNullException(nameof(onReturnFailedSamples));
+            }
+
+            this.serviceClient = serviceClient;
+            this.timeProvider = timeProvider;
+            this.timings = timings;
+            this.onStartCollection = onStartCollection;
+            this.onStopCollection = onStopCollection;
+            this.onSubmitSamples = onSubmitSamples;
+            this.onReturnFailedSamples = onReturnFailedSamples;
+        }
+        
+        public bool IsCollectingData
+        {
+            get
+            {
+                return this.isCollectingData;
+            }
+
+            private set
+            {
+                if (value != this.isCollectingData)
+                {
+                    // state transition
+                    this.ResetLastSuccessful();
+                }
+
+                this.isCollectingData = value;
+            }
+        }
+
+        public TimeSpan UpdateState(string instrumentationKey)
+        {
+            if (string.IsNullOrWhiteSpace(instrumentationKey))
+            {
+                return this.timings.ServicePollingInterval;
+            }
+
+            if (this.firstStateUpdate)
+            {
+                this.ResetLastSuccessful();
+
+                this.firstStateUpdate = false;
+            }
+
+            if (this.IsCollectingData)
+            {
+                // we are currently collecting
+                IList<QuickPulseDataSample> dataSamplesToSubmit = this.onSubmitSamples();
+
+                if (!dataSamplesToSubmit.Any())
+                {
+                    // no samples to submit, do nothing
+                    return this.DetermineBackOffs();
+                }
+
+                bool? keepCollecting = this.serviceClient.SubmitSamples(dataSamplesToSubmit, instrumentationKey);
+
+                switch (keepCollecting)
+                {
+                    case null:
+                        // the request has failed, so we need to return the samples back to the submitter
+                        this.onReturnFailedSamples(dataSamplesToSubmit);
+                        break;
+
+                    case true:
+                        // the service wants us to keep collecting
+                        break;
+
+                    case false:
+                        // the service wants us to stop collection
+                        this.onStopCollection();
+                        break;
+                }
+
+                this.lastSuccessfulSubmit = keepCollecting.HasValue ? this.timeProvider.UtcNow : this.lastSuccessfulSubmit;
```

Comment:
You are taking `this.timeProvider.UtcNow` too often. I think good design would be to take it only twice - once before the call and once after the call. And use those values everywhere for calculation of intervals, passing to the server and so forth. This may eliminate some errors related to precision and accuracy of `DateTime`